### DISCUSSION TITLE
Allow users to define ordering for TestNG listeners

### DISCUSSION THREAD

**Discussion by krmahadevan** [cbeust/testng#2916] [Ideas]
On a high level here's the requirements that we can start off with to decide on the implementation details of https://github.com/testng-team/testng/issues/2874

* User defines a comparator that basically takes in two listener objects (implementations of ITestNGListener)
* Using the comparator the user determines the execution order.
* User plugs in this comparator as a configuration.
* TestNG uses this new comparator to determine order and reverse-order for listeners.

The last portion that needs to be vetted out is to decide if we should include `IReporter` implementations. 

Our `IReporter` api by its design is allowing users to alter the state of the `ITestResult` object (we dont enforce READ-ONLY)

**Comment by juherr** [comment]
Thanks for taking the lead on the feature @krmahadevan 👍 

My points:
* If TestNG must instantiate the comparator, what are the constraints/limitations on the constructor?
* What is the definition of the default order?
* `User plugs in this comparator as a configuration` -> Could you precise your mind?
* I vote to exclude `IReporter` from the feature because it will be more difficult to remove it later if it is a bad idea (but could be easily added if we found a true use case)

**Reply by krmahadevan** [reply]
@juherr 

> If TestNG must instantiate the comparator, what are the constraints/limitations on the constructor?

I was thinking that we will just stick to having the implementation instantiated via the default constructor. It wont be injected with any TestNG objects because I think we need to instantiate this comparator much early and we wont be having any of those objects.

> What is the definition of the default order?

It defaults to the insertion order. In this case, the comparator would be null. 

> User plugs in this comparator as a configuration -> Could you precise your mind?

So a user would basically be doing something like `-comparator org.testng.sample.MyComparator`. Here `org.testng.sample.MyComparator` would be implementing `java.util.Comparator` with its type as `ITestNGMethod`

So it would work similar to how we have the configuration options in our documentation https://testng.org/doc/documentation-main.html#running-testng


> I vote to exclude IReporter from the feature because it will be more difficult to remove it later if it is a bad idea (but could be easily added if we found a true use case)

From a contingency standpoint what you say makes sense. Its easy to expand the scope (by including `IReporter` implementations based on user's ask) rather than try to reduce scope later. So I am fine with this ask. We will exclude `IReporter` impls from this scope.

**Reply by krmahadevan** [reply]
@juherr - Can you please let me know if you are fine with the above discussion points so that I can work on getting this added to the upcoming release. I am realising that there are more and more users who are assuming the listener insertion order and then building on top of TestNG to provide customized experiences which is failing because the order is not predictable

Please see this issue for an example https://github.com/testng-team/testng/issues/3046

**Reply by juherr** [reply]
It looks good.

Why not start by the documentation? We will be able to use it as a specification.

**Reply by krmahadevan** [reply]
> Why not start by the documentation? We will be able to use it as a specification.

You mean, by creating a PR for https://github.com/testng-team/testng-team.github.io that explains how to use this feature?

**Reply by juherr** [reply]
Yes, exactly! WDYT?

**Reply by krmahadevan** [reply]
Sure. Let me work on getting that done over the weekend.

**Reply by krmahadevan** [reply]
@juherr - Please check this PR https://github.com/testng-team/testng-team.github.io/pull/74
